staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/tableconvertor/tableconvertor.go

```diff
@@ -33,50 +31,33 @@ import (
 	"k8s.io/client-go/util/jsonpath"
 )
 
-const printColumnsKey = "x-kubernetes-print-columns"
-
 var swaggerMetadataDescriptions = metav1.ObjectMeta{}.SwaggerDoc()
 
-// New creates a new table convertor for the provided OpenAPI schema. If the printer definition cannot be parsed,
+// New creates a new table convertor for the provided CRD column definition. If the printer definition cannot be parsed,
 // error will be returned along with a default table convertor.
-func New(extensions spec.Extensions) (rest.TableConvertor, error) {
+func New(crdColumns []apiextensions.Column) (rest.TableConvertor, error) {
 	headers := []metav1beta1.TableColumnDefinition{
 		{Name: "Name", Type: "string", Format: "name", Description: swaggerMetadataDescriptions["name"]},
 		{Name: "Created At", Type: "date", Description: swaggerMetadataDescriptions["creationTimestamp"]},
 	}
 	c := &convertor{
 		headers: headers,
 	}
-	format, ok := extensions.GetString(printColumnsKey)
-	if !ok {
-		return c, nil
-	}
-	// "x-kubernetes-print-columns": "custom-columns=NAME:.metadata.name,RSRC:.metadata.resourceVersion"
-	parts := strings.SplitN(format, "=", 2)
-	if len(parts) != 2 || parts[0] != "custom-columns" {
-		return c, fmt.Errorf("unrecognized column definition in 'x-kubernetes-print-columns', only support 'custom-columns=NAME=JSONPATH[,NAME=JSONPATH]'")
-	}
-	columnSpecs := strings.Split(parts[1], ",")
-	var columns []*jsonpath.JSONPath
-	for _, spec := range columnSpecs {
-		parts := strings.SplitN(spec, ":", 2)
-		if len(parts) != 2 || len(parts[0]) == 0 || len(parts[1]) == 0 {
-			return c, fmt.Errorf("unrecognized column definition in 'x-kubernetes-print-columns', must specify NAME=JSONPATH: %s", spec)
-		}
-		path := jsonpath.New(parts[0])
-		if err := path.Parse(parts[1]); err != nil {
-			return c, fmt.Errorf("unrecognized column definition in 'x-kubernetes-print-columns': %v", spec)
+
+	for _, col := range crdColumns {
```

Comment:
I've played a bit with the PR and the only comment that I have is this. If a CRD author specifies columns we should *overwrite* the default ones. This will provide a better UX for CRD authors, otherwise they might end up with unwanted columns. If you look closely [here](https://github.com/kubernetes/kubernetes/blob/1d905bbdfcb07c2660f7c70a1e9d4111285436f1/pkg/printers/internalversion/printers.go) how we specify the columns for the built-in types we always specify full set, there's no defaulting. I'd imagine this being the same for CRDs.

Reply:
Should we default the field to the default columns?

Reply:
I don't think we should allow them to overwrite name.  I'm ok with forcing them to specify createdAt, and that would actually be better because that's an example of a date field we should handle correctly.

Defaulting the field may make sense, not 100% sure.

Reply:
if we enforce name and createAt, I suggest to rename this field to `AdditionalColumns` and don't default it.

Reply:
Or even `AdditionalPrinterColumns`.

Reply:
Let's default and allow them to change it later on. I don't agree with @smarterclayton here about not having the ability to drop `name` field. The author should have that option, and defaulting is exactly the right place for that.